src/roots.jl

```diff
@@ -2,101 +2,67 @@
-
-# Fields:
-  - `initial`: region (as a `Root` object) in which roots are searched.
-  - `contractor`: contractor to use (`Bisection`, `Newton` or `Krawczyk`)
-  - `tol`: tolerance of the search
-"""
-struct DepthFirstSearch{R <: Region, C <: Contractor, T <: Real} <: DepthFirstBBSearch{Root{R}}
-    initial::Root{R}
-    contractor::C
-    tol::T
+struct RootProblem{T}
+    abstol::T
 end
 
-BreadthFirstSearch(X::Region, C::Contractor, tol::Real) = BreadthFirstSearch(Root(X, :unknown), C, tol)
-DepthFirstSearch(X::Region, C::Contractor, tol::Real) = DepthFirstSearch(Root(X, :unknown), C, tol)
-
-root_element(search::BBSearch{Root{R}}) where {R <: Region} = search.initial
-
 function bisect(r::Root)
     Y1, Y2 = bisect(interval(r))
     return Root(Y1, :unknown), Root(Y2, :unknown)
 end
 
-bisect(::BBSearch, r::Root) = bisect(r::Root)
-
-function process(search::Union{BreadthFirstSearch, DepthFirstSearch}, r::Root)
-    contracted_root = search.contractor(r, search.tol)
+function process(contractor, root_problem, r::Root)
+    contracted_root = contractor(r, root_problem.abstol)
     status = root_status(contracted_root)
 
     status == :unique && return :store, contracted_root
-    status == :empty && return :discard, contracted_root
+    status == :empty && return :prune, contracted_root
 
     if status == :unknown
         # Avoid infinite division of intervals with singularity
-        isnan(contracted_root) && diam(r) < search.tol && return :store, r
-        diam(contracted_root) < search.tol && return :store, contracted_root
+        isnan(contracted_root) && diam(r) < root_problem.abstol && return :store, r
+        diam(contracted_root) < root_problem.abstol && return :store, contracted_root
 
-        return :bisect, r
+        return :branch, r
     else
         error("Unrecognized root status: $status")
     end
 end
 
 """
-    branch_and_prune(X, contractor, strategy, tol)
+    branch_and_prune(X, contractor, search_order, tol)
 
 Generic branch and prune routine for finding isolated roots using the given
 contractor to determine the status of a given box `X`.
 
 See the documentation of the `roots` function for explanation of the other
 arguments.
 """
-function branch_and_prune(r::Root, contractor, search, tol)
-    iter = search(r, contractor, tol)
-    local endstate
-    # complete iteration
-    for state in iter
-        endstate = state
-    end
-    return data(endstate)
+function branch_and_prune(r::Root, contractor, search_order, tol)
+    root_problem = RootProblem(tol)
+    search = BranchAndPruneSearch(
+        search_order,
+        X -> process(contractor, root_problem, X),
+        bisect,
+        r
+    )
+    result = bpsearch(search)
+    return vcat(result.final_regions, result.unfinished_regions)
 end
 
 const NewtonLike = Union{Type{Newton}, Type{Krawczyk}}
-const default_strategy = DepthFirstSearch
+const default_search_order = DepthFirst
```

Comment:
out of curiosity, do you have any benchmarks / insights / literature references on how the two search strategies compare for interval root finding? I guess they are pretty much equivalent and there's not a strong advantage of one over the other?

Reply:
As far as I know the main difference is that is you you have a limited number of iteration, going breadth first give you evenly sized intervals over the search domain, while going depth first gives you smallest possible intervals for the subset of the search domain that it had time to go through.